dask/bag/core.py

```diff
@@ -859,26 +862,113 @@ def from_filenames(filenames, chunkbytes=None, encoding=system_encoding):
-        taskss = [_chunk_read_file(fn, chunkbytes, encoding) for fn in full_filenames]
+        taskss = [_chunk_read_file(fn, chunkbytes, encoding, linesep)
+                  for fn in full_filenames]
         d = dict(((name, i), task)
                  for i, task in enumerate(toolz.concat(taskss)))
     else:
         extension = os.path.splitext(filenames[0])[1].strip('.')
         myopen = opens.get(extension, open)
 
-        d = dict(((name, i), (list, (decode_sequence, encoding, (myopen, fn, 'rb'))))
+        d = dict(((name, i), (list, (decode_sequence, encoding,
+                                     (pluck, 0,
+                                      (_readlines, (myopen, fn, 'rb'),
+                                       encoding, linesep)))))
                  for i, fn in enumerate(full_filenames))
 
     return Bag(d, name, len(d))
 
 
-def _chunk_read_file(filename, chunkbytes, encoding):
+def _readlines(fo, encoding, linesep, chunksize=4096):
+    # Get text file encoding.
+    if encoding is None:
+        encoding = getattr(fo, 'encoding', 'utf-8')
+    # Get line separator.
+    if linesep is None:
+        linesep = os.linesep
+    # Get byte representation of the line separator.
+    bin_linesep = get_bin_linesep(encoding, linesep)
+    bin_linesep_len = len(bin_linesep)
+
+    buf = b''
+    fpos = fo.tell()
+    while True:
+        start, end = 0, 0
+        while True:
+            try:
+                pos = buf.index(bin_linesep, start)
+            except ValueError:
+                buf = buf[start:]
+                break
+            else:
+                end = pos + bin_linesep_len
+                yield buf[start:end], fpos + end
+                start = end
+        fpos += end
+        chunk = fo.read(chunksize)
+        if chunk:
+            buf += chunk
+        else:
+            if buf:
+                yield buf, fpos + end + len(buf)
+            break
+
+
+def _textblock(filename, start, end, compression, encoding=None, linesep=None):
+    f = opens.get(compression, open)(filename, 'rb')
+    try:
+        # Get text file encoding.
+        if encoding is None:
+            encoding = getattr(f, 'encoding', 'utf-8')
+        # Get line separator.
+        if linesep is None:
+            linesep = os.linesep
+        # Get byte representation of the line separator.
+        bin_linesep = get_bin_linesep(encoding, linesep)
+        bin_linesep_len = len(bin_linesep)
+
+        # Determine which of the following cases applies:
+        # (1) skip=False: `start` corresponds to the beginning of a line. We
+        #     can start reading lines of this text block.
+        # (2) skip=True: `start` corresponds to a position in the middle of a
+        #     line. We need to skip characters until we reach the beginning of
+        #     the next line.
+        skip = False
+        if start >= bin_linesep_len:
+            f.seek(start - bin_linesep_len)
+            if f.read(bin_linesep_len) != bin_linesep:
```

Comment:
If `bin_linesep_len == 2` how can we be assured that `f.read(bin_linesep_len)` will not split a line separator, e.g. what if we split a windows-style endline across two reads.

```python
>>> f.read(2) 
'A\r'
>>> f.read(2)
'\nB'
```